k8s/k8s.go

```diff
@@ -250,6 +263,118 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, hostname string, calicoCl
+func ipAddrsResult(ipAddrs string, conf utils.NetConf, args *skel.CmdArgs, logger *log.Entry) (*types.Result, error) {
+	var ips []string
+
+	err := json.Unmarshal([]byte(ipAddrs), &ips)
+	if err != nil {
+		logger.WithField("annotation", ipAddrs).Error("Failed to parse ipAddrs as json")
+		return nil, fmt.Errorf("Failed to parse ipAddrs: %s", err)
+	}
+
+	result := types.Result{}
+
+	// annotation value can't be empty.
+	if len(ips) == 0 {
+		logger.WithField("annotation", "cni.projectcalico.org/ipAddrs").Error("No IPs specified")
+		return nil, fmt.Errorf("No IPs specified with annotation \"cni.projectcalico.org/ipAddrs\"")
+	}
+
+	log.Debugf("IPs passed through ipAddrs annotation: %s", ipAddrs)
+
+	// Go through all the IPs passed in as annotation value and call IPAM plugin
+	// for each, and populate the result variable with IP4 and/or IP6 IPs returned
+	// from the IPAM plugin. We also make sure there is only one IPv4 and/or one IPv6
+	// passed in, since CNI spec only supports one of each right now.
+	for _, ip := range ips {
+		ipAddr := net.ParseIP(ip)
+		if ipAddr == nil {
+			logger.WithField("IP", ip).Error("Invalid IP format")
+			return nil, fmt.Errorf("Invalid IP format: %s", err)
+		}
+
+		// It's an IPv6 address if ip.To4 is nil.
+		if ipAddr.To4() == nil {
+			// We only allow one IPv4 and one IPv6 at the moment.
+			// So if we see more than one of IPv4 or IPv6 then we throw an error.
+			// If/when CNI spec supports more than one IP, we can loosen this requirement.
+			if result.IP6 != nil {
+				logger.Error("Can not have more than one IPv6 addresses in ipAddrsNoIpam annotation")
+				return nil, fmt.Errorf("Can not have more than one IPv6 addresses in ipAddrsNoIpam annotation")
+			}
+
+			// Calll setIPArg to set CNI_ARGS and call the IPAM plugin with it, and then unset it.
+			r, err := setIPArg(ipAddr, conf, args, logger)
+			if err != nil {
+				return nil, fmt.Errorf("Error getting IP from IPAM: %s", err)
+			}
+
+			// Set the IP6 part of the result from the type.Result returned by the IPAM plugin.
+			result.IP6 = r.IP6
+			logger.Debugf("Adding IPv6: %s to result", ipAddr.String())
+		} else {
+			// It's an IPv4 address.
+			if result.IP4 != nil {
+				logger.Error("Can not have more than one IPv4 addresses in ipAddrsNoIpam annotation")
+				return nil, fmt.Errorf("Can not have more than one IPv4 addresses in ipAddrsNoIpam annotation")
+			}
+
+			// Calll setIPArg to set CNI_ARGS and call the IPAM plugin with it, and then unset it.
+			r, err := setIPArg(ipAddr, conf, args, logger)
+			if err != nil {
+				return nil, fmt.Errorf("Error getting IP from IPAM: %s", err)
+			}
+
+			// Set the IP4 part of the result from the type.Result returned by the IPAM plugin.
+			result.IP4 = r.IP4
+			logger.Debugf("Adding IPv4: %s to result", ipAddr.String())
+		}
+	}
+
+	return &result, nil
+}
+
+// setIPArg sets CNI_ARGS with the IP requested from the IPAM plugin, then it calls
```

Comment:
nit: the IP is not requested from the IPAM plugin. Rather, it sets CNI_ARGS with the requested IP and then calls the IPAM plugin.